bundles/org.platformio.eclipse.ide.home.ui/src/org/platformio/eclipse/ide/home/internal/ui/handlers/UploadHandler.java

```diff
@@ -0,0 +1,55 @@
+/*******************************************************************************
+ * Copyright (c) 2021 PlatformIO and ArSysOp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Contributors:
+ *     Nikifor Fedorov (ArSysOp) - initial API and implementation
+ *******************************************************************************/
+package org.platformio.eclipse.ide.home.internal.ui.handlers;
+
+import java.nio.file.Paths;
+import java.util.Optional;
+
+import org.eclipse.core.commands.AbstractHandler;
+import org.eclipse.core.commands.ExecutionEvent;
+import org.eclipse.core.commands.ExecutionException;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.core.runtime.ServiceCaller;
+import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.ui.handlers.HandlerUtil;
+import org.platformio.eclipse.ide.home.api.PlatformIO;
+
+public final class UploadHandler extends AbstractHandler {
```

Comment:
I can see 2 more classes that looks very similar. You can have abstract handler for all except call inside `try`